examples/multisig_plain/lib.rs

```diff
@@ -0,0 +1,837 @@
+        #[ink(topic)]
+        from: AccountId,
+    }
+
+    /// Emitted when an owner submits a transaction.
+    #[ink(event)]
+    struct Submission {
+        /// The transaction that was submitted.
+        #[ink(topic)]
+        transaction: TransactionId,
+    }
+
+    /// Emitted when a transaction was canceled.
+    #[ink(event)]
+    struct Cancelation {
+        /// The transaction that was canceled.
+        #[ink(topic)]
+        transaction: TransactionId,
+    }
+
+    /// Emitted when a transaction was executed.
+    #[ink(event)]
+    struct Execution {
+        /// The transaction that was executed.
+        #[ink(topic)]
+        transaction: TransactionId,
+        /// Indicates whether the transaction executed successfully.
+        #[ink(topic)]
+        result: Result<(), ()>,
+    }
+
+    /// Emitted when an owner is added to the wallet.
+    #[ink(event)]
+    struct OwnerAddition {
+        /// The owner that was added.
+        #[ink(topic)]
+        owner: AccountId,
+    }
+
+    /// Emitted when an owner is removed from the wallet.
+    #[ink(event)]
+    struct OwnerRemoval {
+        /// The owner that was removed.
+        #[ink(topic)]
+        owner: AccountId,
+    }
+
+    /// Emitted when the requirement changed.
+    #[ink(event)]
+    struct RequirementChange {
+        /// The new requirement value.
+        new_requirement: u32,
+    }
+
+    impl MultisigPlain {
+        /// The only constructor of the contract.
+        /// A list of owners must be supplied and a number of how many of them must
+        /// confirm a transaction. Duplicate owners are silently dropped.
+        #[ink(constructor)]
+        fn new(&mut self, owners: Vec<AccountId>, requirement: u32) {
+            for owner in &owners {
+                self.is_owner.insert(*owner, ());
+                self.owners.push(*owner);
+            }
+            self.ensure_requirement_is_valid(self.owners.len(), requirement);
+            assert!(self.is_owner.len() == self.owners.len());
+            self.requirement.set(requirement);
+        }
+
+        /// Add a new owner to the contract.
+        /// Panics if the owner already exists.
+        /// Only callable by the wallet itself.
+        #[ink(message)]
+        fn add_owner(&mut self, new_owner: AccountId) {
+            self.ensure_from_wallet();
+            self.ensure_no_owner(&new_owner);
+            self.ensure_requirement_is_valid(self.owners.len() + 1, *self.requirement);
+            self.is_owner.insert(new_owner, ());
+            self.owners.push(new_owner);
+            self.env().emit_event(OwnerAddition { owner: new_owner });
+        }
+
+        /// Remove an owner from the contract.
+        /// Only callable by the wallet itself. If by doing this the amount of owners
+        /// would be smaller than the requirement it is adjusted to be exactly the
+        /// number of owners. Panics if `owner` is no owner of the wallet.
```

Comment:
Is this the semantics of the original multi-sig because I kinda dislike this implied effect but maybe I just don't know why this is useful to have. I'd generally just trap (panic) in such cases. If it is the original semantics we might just keep it as it is though.

Reply:
Yes this is the gnosis semantic. But we can change it. We changed other things already.

Reply:
I think we should leave it the way it is for the following reason: If we introduce the panic behavior you need to lower the requirement before removing an owner. This introduces a race where after lowering the requirement a transaction can get confirmed by too few people (including the owner you want to remove).